File: core/src/main/java/dev/bee/kanjianki/core/SimilarKanjiStorageKeys.java
package dev.bee.kanjianki.core;

public final class SimilarKanjiStorageKeys {
    private static final String PAIR_DELIMITER = "\u0000";
    private static final String CHOICE_DELIMITER = "\u0001";

    private SimilarKanjiStorageKeys() {
    }

    public static String[] canonicalPair(String first, String second) {
        if (first.compareTo(second) <= 0) {
            return new String[]{first, second};
        }
        return new String[]{second, first};
    }

    public static String pairKey(String first, String second, String source) {
        return first + PAIR_DELIMITER + second + PAIR_DELIMITER + source;
    }

    public static String choiceKey(String targetKanji, String choiceSignature) {
        return targetKanji + CHOICE_DELIMITER + (choiceSignature == null ? "" : choiceSignature);
    }

    public static String[] splitChoiceKey(String key) {
        if (key == null) {
            return new String[0];
        }
        return key.split(CHOICE_DELIMITER, 2);
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/SimilarKanjiStorageKeys.kt
package dev.bee.kanjianki.core

object SimilarKanjiStorageKeys {
    private const val PAIR_DELIMITER = "\u0000"
    private const val CHOICE_DELIMITER = "\u0001"

    @JvmStatic
    fun canonicalPair(first: String?, second: String?): Array<String> {
        val safeFirst = first!!
        val safeSecond = second!!
        if (safeFirst <= safeSecond) {
            return arrayOf(safeFirst, safeSecond)
        }
        return arrayOf(safeSecond, safeFirst)
    }

    @JvmStatic
    fun pairKey(first: String?, second: String?, source: String?): String {
        return first.toString() + PAIR_DELIMITER + second + PAIR_DELIMITER + source
    }

    @JvmStatic
    fun choiceKey(targetKanji: String?, choiceSignature: String?): String {
        return targetKanji.toString() + CHOICE_DELIMITER + (choiceSignature ?: "")
    }

    @JvmStatic
    fun splitChoiceKey(key: String?): Array<String> {
        if (key == null) {
            return emptyArray()
        }
        return key.split(CHOICE_DELIMITER, limit = 2).toTypedArray()
    }
}
